config/module/registry.go

```diff
@@ -0,0 +1,185 @@
+			Path:   defaultApiPath,
+		}
+	}
+
+	service := regURL.String()
+
+	if service[len(service)-1] != '/' {
+		service += "/"
+	}
+
+	return service
+}
+
+// Lookup module versions in the registry.
+func lookupModuleVersions(d *disco.Disco, module *regsrc.Module) (*response.ModuleVersions, error) {
+	service := discoverRegURL(d, module)
+
+	location := fmt.Sprintf("%s%s/versions", service, module.Module())
+	log.Printf("[DEBUG] fetching module versions from %q", location)
+
+	req, err := http.NewRequest("GET", location, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set(xTerraformVersion, tfVersion)
+
+	if d == nil {
+		d = regDisco
+	}
+
+	// if discovery required a custom transport, then we should use that too
+	client := httpClient
+	if d.Transport != nil {
+		client = &http.Client{
+			Transport: d.Transport,
+			Timeout:   requestTimeout,
+		}
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// OK
+	case http.StatusNotFound:
+		return nil, errModuleNotFound(module.String())
+	default:
+		return nil, fmt.Errorf("error looking up module versions: %s", resp.Status)
+	}
+
+	var versions response.ModuleVersions
+
+	dec := json.NewDecoder(resp.Body)
+	if err := dec.Decode(&versions); err != nil {
+		return nil, err
+	}
+
+	return &versions, nil
+}
+
+// lookup the location of a specific module version in the registry
+func lookupModuleLocation(d *disco.Disco, module *regsrc.Module, version string) (string, error) {
+	service := discoverRegURL(d, module)
+
+	var download string
+	if version == "" {
+		download = fmt.Sprintf("%s%s/download", service, module.Module())
+	} else {
+		download = fmt.Sprintf("%s%s/%s/download", service, module.Module(), version)
+	}
+
+	log.Printf("[DEBUG] looking up module location from %q", download)
+
+	req, err := http.NewRequest("GET", download, nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set(xTerraformVersion, tfVersion)
+
+	// if discovery required a custom transport, then we should use that too
+	client := httpClient
+	if regDisco.Transport != nil {
+		client = &http.Client{
+			Transport: regDisco.Transport,
+			Timeout:   requestTimeout,
+		}
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	// there should be no body, but save it for logging
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("error reading response body from registry: %s", err)
+	}
+
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusNoContent:
+		// OK
+	case http.StatusNotFound:
+		return "", fmt.Errorf("module %q version %q not found", module, version)
+	default:
+		// anything else is an error:
+		return "", fmt.Errorf("error getting download location for %q: %s resp:%s", module, resp.Status, body)
+	}
+
+	// the download location is in the X-Terraform-Get header
+	location := resp.Header.Get(xTerraformGet)
```

Comment:
My understanding of the protocol defined in the RFC was that `X-Terraform-Get` here is just one possible response and that the server is also free to return a redirect or to just return the archive directly. I think the expectation was that this "download" URL would be passed directly to `go-getter` and then it would do all of the things it usually does to try to sniff out the appropriate behavior.

That does throw a bit of a wrench in the works for auth though, since I don't think we can add the necessary `Authorization` header to `go-getter`'s request with today's interface, unless we were to do something interesting with a custom getter.

Maybe this is okay for now, but I'm nervous that this is more constrained than the protocol intended and we may grow to regret this if we need to change our implementation in future.

Reply:
Oh, I must have missed that. I thought the header was the only method for locating the resource at the moment.